// noomman/noomman/ClassModel.js


var database = require('./database');
const InstanceSet = require('./InstanceSet');
const Instance = require('./Instance');
const Attribute = require('./Attribute');
const Relationship = require('./Relationship');
const Diffable = require('./Diffable');
const SuperSet = require('./SuperSet');
const NoommanErrors = require('./NoommanErrors');

const AllClassModels = [];

/*
 * Class ClassModel
 * A class which defines the schema for a Class which will be stored in the mongo database.
 *    Has methods for querying the underlying database collections for instances of a ClassModel
 *    and its sub ClassModels. Enables inheritance of attributes, relationships, methods, etc.
 */
class ClassModel {

    /*
     * constructor(schema)
     * Creates an instance of ClassModel with the given schema.
     * Parameters: 
     * - schema - Object - A schema describing the properties of this ClassModel.
     *    {
     *       className: String (required),
     *       superClasses: [ ClassModel ],
     *       useSuperClassCollection: Boolean,
     *       abstract: Boolean,
     *       auditable: Boolean,
     *       attributes: [
     *          {
     *             name: String (required), 
     *             type: String (required), 
     *             list: Boolean,
     *             required: Boolean,
     *             unique: Boolean,
     *             sensitive: Boolean,
     *             mutex: String,
     *             requiredGroup: String,
     *          },
     *       ],
     *       relationships: [
     *          {
     *             name: String (required),
     *             toClass: String (required),
     *             singular: Boolean (required),
     *             required: Boolean,
     *             owns: Boolean,
     *             mirrorRelationship: String,
     *             mutex: String,
     *             requiredGroup: String,
     *          }
     *       ],
     *       crudFunctions: {
     *          createControl: Function,
     *          readControl: Function,
     *          updateControl: Function,
     *          deleteControl: Function,
     *          sensitiveControl: Function,
     *       },
     *       validations: [ Function ],
     *       indices: [ fieldOrSpec ], 
     *       staticMethods: {
     *          String: Function,
     *       },
     *       nonStaticMethods: {
     *          String: Function,
     *       },
     *    }
     * Returns
     * - ClassModel - The ClassModel created according to the given schema.
     * Throws
     * - NoommanConstructorError - If constructorValidations() method throws a NoommanConstructorError.
     */
    constructor(schema) {
        this.constructorValidations(schema);

        this.className = schema.className;
        this.subClasses = [];
        this.abstract = schema.abstract;
        this.useSuperClassCollection = schema.useSuperClassCollection;
        this.superClasses = schema.superClasses ? schema.superClasses : [];
        this.indices = schema.indices ? schema.indices : [];
        this.collection = schema.useSuperClassCollection ? schema.superClasses[0].collection : schema.className.toLowerCase();
        this.auditable = schema.auditable === true;
        this.staticMethods = schema.staticMethods ? schema.staticMethods : {};
        this.nonStaticMethods = schema.nonStaticMethods ? schema.nonStaticMethods : {};

        if (this.superClasses.length === 0 && this.className !== 'NoommanClassModel') {
            this.superClasses.push(NoommanClassModel);
        }

        if (!schema.useSuperClassCollection) {
            const lastLetter = schema.className.substr(-1).toLowerCase();
            this.collection = schema.className.toLowerCase();
            if (lastLetter === 's') {
                this.collection = this.collection + 'e';
            }
            this.collection = this.collection + 's';
        }
        
        this.attributes = [];
        this.relationships = [];

        if (schema.attributes) {
            for (const attribute of schema.attributes) {
                this.attributes.push(new Attribute(attribute));
            }
        }

        if (schema.relationships) {
            for (const relationship of schema.relationships) {
                this.relationships.push(new Relationship(relationship));
            }
        }

        this.createControlMethods = [];
        this.readControlMethods = [];
        this.updateControlMethods = [];
        this.deleteControlMethods = [];
        this.sensitiveControlMethods = [];

        if (schema.crudControls) {
            if (schema.crudControls.createControl) {
                this.createControlMethods.push(schema.crudControls.createControl);
            }
            if (schema.crudControls.readControl) {
                this.readControlMethods.push(schema.crudControls.readControl);
            }
            if (schema.crudControls.updateControl) {
                this.updateControlMethods.push(schema.crudControls.updateControl);
            }
            if (schema.crudControls.deleteControl) {
                this.deleteControlMethods.push(schema.crudControls.deleteControl);
            }
            if (schema.crudControls.sensitiveControl) {
                this.sensitiveControlMethods.push(schema.crudControls.sensitiveControl);
            }
        }

        this.validations = [];

        if (schema.validations) {
            for (const validation of schema.validations) {
                this.validations.push(validation);
            }
        }

        if (schema.superClasses) {
            for (const superClass of schema.superClasses) {
                this.attributes = this.attributes.concat(superClass.attributes);
                this.relationships = this.relationships.concat(superClass.relationships);
                this.indices = this.indices.concat(superClass.indices);
                this.createControlMethods = this.createControlMethods.concat(superClass.createControlMethods);
                this.readControlMethods = this.readControlMethods.concat(superClass.readControlMethods);
                this.updateControlMethods = this.updateControlMethods.concat(superClass.updateControlMethods);
                this.deleteControlMethods = this.deleteControlMethods.concat(superClass.deleteControlMethods);
                this.sensitiveControlMethods = this.sensitiveControlMethods.concat(superClass.sensitiveControlMethods);
                this.validations = this.validations.concat(superClass.validations);
                this.auditable = superClass.auditable ? true : this.auditable;
                this.inheritStaticMethods(superClass);
                this.inheritNonStaticMethods(superClass);
                superClass.subClasses.push(this);
            }
        }

        for (const staticMethod of Object.keys(this.staticMethods)) {
            this[staticMethod] = this.staticMethods[staticMethod];
        }

        for (const nonStaticMethod of Object.keys(this.nonStaticMethods)) {
            this[nonStaticMethod] = this.nonStaticMethods[nonStaticMethod];
        }

        AllClassModels[this.className] = this;
    }

    /*
     * constructorValidations(schema)
     * Throws an error if the schema passed to constructor() is not of expected type or 
     *    is functionally invalid.
     * Parameters: 
     * - schema - Object - A schema describing the properties of this ClassModel. 
     *    See schema parameter or constructor() method.
     * Throws
     * - NoommanConstructorError - If parameterShapeConstructorValidations throws a NoommanConstructorError.
     * - NoommanConstructorError - If crudControlsConstructorValidations throws a NoommanConstructorError.
     * - NoommanConstructorError - If sensitiveAttributesContructorValidations throws a NoommanConstructorError.
     * - NoommanConstructorError - If customMethodsContructorValidations throws a NoommanConstructorError.
     * - NoommanConstructorError - If inheritanceConstructorValidations throws a NoommanConstructorError.
     */
    constructorValidations(schema) {
        ClassModel.paramterShapeConstructorValidations(schema);

        ClassModel.crudControlsConstructorValidations(schema);

        ClassModel.sensitiveAttributesContructorValidations(schema);

        ClassModel.customMethodsContructorValidations(schema);

        ClassModel.inheritanceConstructorValidations(schema);
    }

    /*
     * paramterShapeConstructorValidations(schema)
     * Validates that the schema passed to constructor() is the correct shape and property types.
     * Parameters
     * - schema - Object - See constructor parameter definition.
     * Throws
     * - NoommanConstructorError - If property className is omitted.
     * - NoommanConstructorError - If property attributes is provided and is not an Array.
     * - NoommanConstructorError - If property relationships is provided and is not an Array.
     * - NoommanConstructorError - If property superClasses is provided and is not an Array.
     * - NoommanConstructorError - If property superClasses is provided and is an empty Array.
     * - NoommanConstructorError - If property useSuperClassCollection is true and superClasses
     *    is omitted or contains more than one ClassModel.
     * - NoommanConstructorError - If property auditable is provided and is not a Boolean.
     * - NoommanConstructorError - If property indices is provided and is not an Array.
     * - NoommanConstructorError - If property validations is provided and is not an Array.
     */
    static paramterShapeConstructorValidations(schema) {        
        if (!schema.className)
            throw new NoommanErrors.NoommanConstructorError('className is required.');

        if (schema.attributes && !Array.isArray(schema.attributes))
            throw new NoommanErrors.NoommanConstructorError('If attributes is set, it must be an Array.');

        if (schema.relationships && !Array.isArray(schema.relationships))
            throw new NoommanErrors.NoommanConstructorError('If relationships is set, it must be an Array.');

        if (schema.superClasses && !Array.isArray(schema.superClasses))
            throw new NoommanErrors.NoommanConstructorError('If superClasses is set, it must be an Array.');

        if (schema.superClasses && schema.superClasses.length == 0)
            throw new NoommanErrors.NoommanConstructorError('If superClasses is set, it cannot be an empty Array.');

        if (schema.useSuperClassCollection && (!schema.superClasses || schema.superClasses.length !== 1)) {
            throw new NoommanErrors.NoommanConstructorError('If useSuperClassCollection is true, a single super class must be provided.');
        }

        if (schema.auditable !== undefined && typeof(schema.auditable) !== 'boolean') {
            throw new NoommanErrors.NoommanConstructorError('If auditable is provided, it must be a boolean.');
        }

        if (schema.indices !== undefined && !Array.isArray(schema.indices)) {
            throw new NoommanErrors.NoommanConstructorError('If indices are provided, indices must be an array.');
        }

        if (schema.validations && !Array.isArray(schema.validations))
            throw new NoommanErrors.NoommanConstructorError('If validations are provided, it must be an Array.');
    }

    /*
     * crudControlsConstructorValidations(schema)
     * If the property crudControls is provided and any of the properties are not a Function, will
     *    throw an Error.
     * Parameters
     * - schema - Object - See constructor parameter definition.
     * Throws
     * - NoommanConstructorError - If property createControl of property crudControls is not a Function.
     * - NoommanConstructorError - If property readControl of property crudControls is not a Function.
     * - NoommanConstructorError - If property updateControl of property crudControls is not a Function.
     * - NoommanConstructorError - If property deleteControl of property crudControls is not a Function.
     * - NoommanConstructorError - If property sensitiveControl of property crudControls is not a Function.
     */
    static crudControlsConstructorValidations(schema) {
        if (schema.crudControls) {
            if (schema.crudControls.readControl && typeof(schema.crudControls.readControl) !== 'function') {
                throw new NoommanErrors.NoommanConstructorError('If a readControl method is provided, it must be a function.');
            }
            if (schema.crudControls.createControl && typeof(schema.crudControls.createControl) !== 'function') {
                throw new NoommanErrors.NoommanConstructorError('If a createControl method is provided, it must be a function.');
            }
            if (schema.crudControls.updateControl && typeof(schema.crudControls.updateControl) !== 'function') {
                throw new NoommanErrors.NoommanConstructorError('If a updateControl method is provided, it must be a function.');
            }
            if (schema.crudControls.deleteControl && typeof(schema.crudControls.deleteControl) !== 'function') {
                throw new NoommanErrors.NoommanConstructorError('If a deleteControl method is provided, it must be a function.');
            }
            if (schema.crudControls.sensitiveControl && typeof(schema.crudControls.sensitiveControl) !== 'function') {
                throw new NoommanErrors.NoommanConstructorError('If a sensitiveControl method is provided, it must be a function.');
            }
        }
    }

    /*
     * sensitiveAttributesContructorValidations(schema)
     * Will throw an error if an attribute in ClassModel schema or parent ClassModel schema has an attribute
     *    marked sensitive, and this ClassModel does not ahve a sensitiveControl method, and vice versa.
     * Parameters
     * - schema - Object - See constructor parameter definition.
     * Throws
     * - NoommanConstructorError - If ClassModel has a sensitive attribute but no sensitiveControl method.
     * - NoommanConstructorError - If ClassModel has a sensitiveControl method but no sensitive attribute.
     */
    static sensitiveAttributesContructorValidations(schema) {
        let allAttributes = [];

        if (schema.superClasses) {
            for (const superClass of schema.superClasses) {
                allAttributes = allAttributes.concat(superClass.attributes);
            }
        }

        if (schema.attributes) {
            allAttributes = allAttributes.concat(schema.attributes)
        }

        if (allAttributes.length) {        
            let sensitiveAttributes = false;
            for (const attribute of allAttributes) {
                if (attribute.sensitive === true) {
                    sensitiveAttributes = true;
                    break;
                }
            }

            if (sensitiveAttributes) {
                if (!schema.crudControls || !schema.crudControls.sensitiveControl) {
                    throw new NoommanErrors.NoommanConstructorError('At least one attribute is marked sensitive, but no sensitiveControl method is provided.');
                }
            }
            else {
                if (schema.crudControls && schema.crudControls.sensitiveControl) {
                    throw new NoommanErrors.NoommanConstructorError('A sensitiveControl method was provided, but no attributes are marked sensitive.');
                }
            }
        }
        else {
            if (schema.crudControls && schema.crudControls.sensitiveControl) {
                throw new NoommanErrors.NoommanConstructorError('A sensitiveControl method was provided, but no attributes are marked sensitive.');
            }
        }

    }

    /*
     * customMethodsContructorValidations(schema)
     * Will throw an error if staticMethods and nonStaticMethods properties of schema are not Objects 
     *    (if provided) or if any of their properties are not functions.
     * Parameters
     * - schema - Object - See constructor parameter definition.
     * Throws
     * - NoommanConstructorError - If property staticMethods is provided but is not an Object.
     * - NoommanConstructorError - If property staticMethods is provided, and any of its properties is not a Function.
     * - NoommanConstructorError - If property staticMethods is provided, and any of its properties has the same
     *    name as a built in noomman method on class ClassModel.
     * - NoommanConstructorError - If property nonStaticMethods is provided but is not an Object.
     * - NoommanConstructorError - If property nonStaticMethods is provided, and any of its properties is not a Function.
     * - NoommanConstructorError - If property nonStaticMethods is provided, and any of its properties has the same
     *    name as a built in noomman method on class Instance.
     */
    static customMethodsContructorValidations(schema) {
        if (schema.staticMethods !== undefined) {
            if (typeof(schema.staticMethods) !== 'object') {
                throw new NoommanErrors.NoommanConstructorError('If staticMethods is provided, it must be an object.');
            }

            for (const staticMethod in schema.staticMethods) {
                if (typeof(schema.staticMethods[staticMethod]) !== 'function') {
                    throw new NoommanErrors.NoommanConstructorError('Each property of staticMethods object must be a function.');
                }

                if (Object.getOwnPropertyNames(ClassModel.prototype).includes(staticMethod)) {
                    throw new NoommanErrors.NoommanConstructorError('Attempt to add a static method with the same name as a built in Noomman method: ' + staticMethod + '.');
                }
            }
        }

        if (schema.nonStaticMethods !== undefined) {
            if (typeof(schema.nonStaticMethods) !== 'object') {
                throw new NoommanErrors.NoommanConstructorError('If nonStaticMethods is provided, it must be an object.');
            }

            for (const nonStaticMethod in schema.nonStaticMethods) {
                if (typeof(schema.nonStaticMethods[nonStaticMethod]) !== 'function') {
                    throw new NoommanErrors.NoommanConstructorError('Each property of nonStaticMethods object must be a function.');
                }

                if (Object.getOwnPropertyNames(Instance.prototype).includes(nonStaticMethod)) {
                    throw new NoommanErrors.NoommanConstructorError('Attempt to add a non-static method with the same name as a built in Noomman method: ' + nonStaticMethod + '.');
                }
            }
        }
    }

    /*
     * inheritanceConstructorValidations(schema)
     * Validates that inheritance related portions of the schema are of the correct types and
     *    do not logically conflict with one another.
     * Parameters
     * - schema - Object - See constructor parameter definition.
     * Throws
     * - NoommanConstructorError - If properties useSuperClassCollection and abstract are both true.
     * - NoommanConstructorError - If property superClasses is provided, and any super ClassModel has an attribute
     *    with the same name as a attribute defined in the attributes property of the schema.
     * - NoommanConstructorError - If property superClasses is provided, and any super ClassModel has an relationship
     *    with the same name as a relationship defined in the relationships property of the schema.
     * - NoommanConstructorError - If any ClassModel in the superClasses property has useSuperClassModel set to true.
     * - NoommanConstructorError - If property auditable is false, but a ClassModel in the superClasses property has
     *    auditable set to true.
     * - NoommanConstructorError - If properties abstract and useSuperClassCollection are both true.
     */
    static inheritanceConstructorValidations(schema) {
        if (schema.useSuperClassCollection && schema.abstract) {
            throw new NoommanErrors.NoommanConstructorError('If useSuperClassCollection is true, abstract cannot be true.')
        }

        if (schema.superClasses) {
            for (const superClass of schema.superClasses) {

                for (const attribute of superClass.attributes) {
                    if (schema.attributes && schema.attributes.map(attribute => attribute.name).includes(attribute.name)) {
                        throw new NoommanErrors.NoommanConstructorError('Sub class schema cannot contain the same attribute names as a super class schema.');
                    }
                }
                for (const relationship of superClass.relationships) {
                    if (schema.relationships && schema.relationships.map(relationship => relationship.name).includes(relationship.name)) {
                        throw new NoommanErrors.NoommanConstructorError('Sub class schema cannot contain the same relationship names as a super class schema.');
                    }
                }

                if (superClass.useSuperClassCollection) {
                    throw new NoommanErrors.NoommanConstructorError('You cannot create a sub class of a class which has useSuperClassCollection set to true.');
                }

                if (schema.auditable === false && superClass.auditable === true) {
                    throw new NoommanErrors.NoommanConstructorError('You cannot create a non-auditable sub class of an auditable super class.');
                }
            }
        } 

        if (schema.useSuperClassCollection && schema.abstract) 
            throw new NoommanErrors.NoommanConstructorError('If useSuperClassCollection is true, the class cannot be abstract.');
    }

    /* 
     * finalize()
     * For each defined ClassModel, runs post-constructor validations and applies indices.
     *    Run only after ALL class models have been created.
     * Returns
     * - Promise<undefined> - A Promise which resolves to undefined if successful.
     * Throws
     * - NoommanClassModelError - Validations errors thrown by validateRelationships()
     */
    static async finalize() {
        for (const classModel of AllClassModels) { 
            classModel.validateRelationships();
        }
        for (const classModel of AllClassModels) {
            await classModel.index();
        }
    }

    /* 
     * index()
     * Adds any user defined or noomman automatic indices to the collection for this ClassModel.
     *    This method is called automatically as part of the finalize() static method.
     * Returns
     * - Promise<undefined> - A Promise which resolves to undefined if indexing is successful.
     */
    async index() {
        const indicesApplied = [];

        for (const index of this.indices) {
            indicesApplied.push(await database.index(this.collection, index));
        }

        if (this.useSuperClassCollection) {
            indicesApplied.push(await database.index(this.collection, '__t'));
        }

        return indicesApplied;
    }

    /* 
     * validateRelationships()
     * Determines if all the relationships defined on this ClassModel are valid.
     *    This must be called only after all the ClassModels have been definied, as it checks
     *    that the toClass of each relationship is a defined ClassModel and that two way 
     *    relationships are correct on both sides of the relationship.
     * Throws 
     * - NoommanClassModelError - If no ClassModel exists with the value of a relationship's toClass property.
     * - NoommanClassModelError - If a two-way relationship is defined and the related ClassModel does not have
     *    a relationship a name matching the mirrorRelationship property of the relationship.
     * - NoommanClassModelError - If a two-way relationship is defined but the mirrorRelationship properties of 
     *    the two relationships are not the name of the other relationship.
     */
    validateRelationships() {
        for (const relationship of this.relationships) {
            const toClass = AllClassModels[relationship.toClass];
            if (toClass === undefined) {
                throw new NoommanErrors.NoommanClassModelError(
                    'Relationship ' + this.className + '.' + relationship.name + 
                    ' is a reference to a Class Model that does not exist: ' + relationship.toClass + '.'
                );
            }

            if (relationship.mirrorRelationship !== undefined) {
                let mirrorRelationship = toClass.relationships.filter(r => r.name === relationship.mirrorRelationship);
                if (mirrorRelationship.length === 0) {
                    throw new NoommanErrors.NoommanClassModelError('Invalid two-way relationship. ' + 
                        this.className + '.' + relationship.name + ' is missing mirror relationship ' + 
                        toClass.className + '.' + relationship.mirrorRelationship + '.'
                    );
                }
                mirrorRelationship = mirrorRelationship[0];

                if (mirrorRelationship.toClass !== this.className) {
                    throw new NoommanErrors.NoommanClassModelError('Invalid two-way relationship. ' + 
                        this.className + '.' + relationship.name + '. Mirror relationship ' + 
                        toClass.className + '.' + relationship.mirrorRelationship + 
                        ' has incorrect toClass: ' + mirrorRelationship.toClass + '.'
                    );
                }

                if (mirrorRelationship.mirrorRelationship !== relationship.name) {
                    throw new NoommanErrors.NoommanClassModelError('Invalid two-way relationship. ' + 
                        this.className + '.' + relationship.name + '. Mirror relationship ' + 
                        toClass.className + '.' + relationship.mirrorRelationship + 
                        ' has incorrect mirrorRelationship: ' + mirrorRelationship.mirrorRelationship + '.'
                    );
                }
            }
        }
    }

    /*
     * inheritStaticMethods(fromClass) 
     * Adds all staticMethods on the fromClass paramter to this ClassModel. Called by constructor(). 
     * Paramters
     * - fromClass - ClassModel - The classModel to inherit methods from.
     */
    inheritStaticMethods(fromClass) {
        for (const staticMethod of Object.keys(fromClass.staticMethods)) {
            if (!Object.keys(this.staticMethods).includes(staticMethod)) {
                this.staticMethods[staticMethod] = fromClass.staticMethods[staticMethod];
            }
        }
    }

    /* 
     * inheritNonStaticMethods(fromClass) 
     * Adds all monStaticMethods on the fromClass paramter to this ClassModel. Called by constructor(). 
     * Paramters
     * - fromClass - ClassModel - The classModel to inherit methods from.
     */
    inheritNonStaticMethods(fromClass) {
        for (const nonStaticMethod of Object.keys(fromClass.nonStaticMethods)) {
            if (!Object.keys(this.nonStaticMethods).includes(nonStaticMethod)) {
                this.nonStaticMethods[nonStaticMethod] = fromClass.nonStaticMethods[nonStaticMethod];
            }
        }
    }

    /* 
     * toString()
     * Returns 
     * - String - A string with this ClassModel's className, followed by a new line.
     */
    toString() {
        return this.className + '\n';
    }

    /* 
     * isInstanceOfThisClass(instance)
     * Determines if the given instance is an instance of this ClassModel or any of 
     *    this ClassModels children.
     * Parameters
     * - instance - Instance - An instance of noomman class Instance
     * Returns
     * - Boolean - True if the given Instance is an Instance for this ClassModel or its children, false otherwise.
     */
    isInstanceOfThisClass(instance) {
        if (instance.classModel === this)
            return true;

        return instance.classModel.allSuperClasses().map(c => c.className).includes(this.className);
    }

    /* 
     * isInstanceSetOfThisClass(instanceSet)
     * Determines if the given instanceSet is an InstanceSet of this ClassModel or any of 
     *    this ClassModel's children.
     * Parameters
     * - instanceSet - InstanceSet - An instance of noomman class InstanceSet
     * Returns
     * - Boolean - True if the given InstanceSet is an InstanceSet of this ClassModel or its children, false otherwise.
     */
    isInstanceSetOfThisClass(instanceSet) {
        if (instanceSet.classModel === this)
            return true;

        return instanceSet.classModel.allSuperClasses().map(c => c.className).includes(this.className);
    }

    /* 
     * getRelatedClassModel(relationshipName)
     * Retreives the ClassModel for the given relationshipName, corresponding to a relationship on this 
     *    ClassModel, from the internal static property AllCLassModels.
     * Parameters
     * - relationshipName - String - A string matching the name property of one of the relationships of 
     *    this ClassModel.
     * Returns
     * - ClassModel - The ClassModel with the className matching the toClass of the relationship with name
     *    matching the relationshipName parameter, defined for this ClassModel (or a parent). 
     */
    getRelatedClassModel(relationshipName) {
        return AllClassModels[this.relationships.filter(relationship => relationship.name === relationshipName)[0].toClass];
    }

    /* 
     * getClassModel(className)
     * Retreives a ClassModel with the given name from the internal static property AllClassModels.
     * Parameters
     * - className - String - A string which should match the className property of the ClassModel one 
     *    wishes to retrieve.
     * Returns
     * - ClassModel - The ClassModel whose className property matches the given className.
     */
    static getClassModel(className) {
        return AllClassModels[className];
    }

    /* 
     * getAllClassModelNames()
     * Retrieves all the classNames of all created ClassModels, except NoommanClassModel.
     * Returns
     * - Array<String> - the classNames of all created ClassModels, except NoommanClassModel.
     */

    static getAllClassModelNames() {
        const names = [];
        for (const index in AllClassModels) {
            const classModel = AllClassModels[index];
            if (classModel.className !== 'NoommanClassModel') {
                names.push(classModel.className);
            }
        }
        
        return names;
    }

    /* 
     * validateAttribute(attributeName, value)
     * Validates that the given value is a valid value for the Attribute with the given attributeName.
     *    Calls method attribute.validate().
     * Parameters
     * - attributeName - String - The name of an attribute of this ClassModel to validate against.
     * - value - Any - A value to validate against the attribute of this ClassModel with the given name.
     * Throws
     * - NoommanValidationError - If the value is an invalid value for the attribute of this ClassModel with the given name. 
     *    See errors on method attribute.validate().
     */ 
    validateAttribute(attributeName, value) {
        const attribute = this.attributes.filter(attribute => attribute.name === attributeName);

        if (attribute.length === 0)
            throw new NoommanErrors.NoommanValidationError('classModel.validateAttribute() called with an invalid attribute name.');

        attribute[0].validate(value);
    }

    /* 
     * valueValidForSingularRelationship(value, relationshipName)
     * Determines if the given value is a valid value for the singular relationship on this ClassModel
     *    with the given relationshipName. Value is considered valid if it is an Instance of the ClassModel
     *    that has the same name as the toClass property of the reltaionship on this ClassModel matching
     *    the given relationshipName (or any sub-ClassModel thereof). Null is also a valid value.
     * Parameters
     * - value - Any - A value to validate.
     * - relationshipName - String - The name of a singular relationship on this ClassModel.
     * Returns
     * - Boolean - True if value is valid for the relationship on this ClassModel matching relationshipName,
     *    false otherwise.
     * Throws
     * - NoommanValidationError - If relationshipName does not match the name property of a singular relationship defined 
     *    for this ClassModel.
     */
    valueValidForSingularRelationship(value, relationshipName) {
        const relationship = this.relationships.filter(relationship => relationship.name === relationshipName && relationship.singular);
        if (relationship.length === 0)
            throw new NoommanErrors.NoommanValidationError('classModel.valueValidForSingularRelationship() called with an invalid relationship name.');

        const toClass = AllClassModels[relationship[0].toClass];

        if (value === null)
            return true;

        if (!(value instanceof Instance))
            return false;

        if (!toClass.isInstanceOfThisClass(value))
            return false;

        return true;
    }

    /* 
     * valueValidForNonSingularRelationship(value, relationshipName)
     * Determines if the given value is a valid value for the non-singular relationship on this ClassModel
     *    with the given relationshipName. Value is considered valid if it is an InstanceSet of the ClassModel
     *    that has the same name as the toClass property of the reltaionship on this ClassModel matching
     *    the given relationshipName (or any sub-ClassModel thereof). Null is also a valid value.
     * Parameters
     * - value - Any - A value to validate.
     * - relationshipName - String - The name of a non-singular relationship on this ClassModel.
     * Returns
     * - Boolean - True if value is valid for the relationship on this ClassModel matching relationshipName,
     *    false otherwise.
     * Throws
     * - NoommanValidationError - If relationshipName does not match the name property of a non-singular relationship defined 
     *    for this ClassModel.
     */
    valueValidForNonSingularRelationship(value, relationshipName) {
        const relationship = this.relationships.filter(relationship => relationship.name === relationshipName && !relationship.singular);
        if (relationship.length === 0)
            throw new NoommanErrors.NoommanValidationError('classModel.valueValidForNonSingularRelationship() called with an invalid relationship name.');

        const toClass = AllClassModels[relationship[0].toClass];

        if (value === null)
            return true;

        if (!(value instanceof InstanceSet))
            return false;

        if (!toClass.isInstanceSetOfThisClass(value))
            return false;

        return true;
    }

    /* 
     * cardinalityOfRelationship(relationshipName)
     * Returns an object representing the cardinality of the relationship on this ClassModel matching the
     *    given relationshipName.
     * Parameters
     * - relationshipName - String - A string matching the name property of a relationship defined for this
     *    ClassModel
     * Returns
     * - Object - An object with two properties, to and from. 
     * {
     *    to: String - Either '1', or 'many',
     *    from: String - Either null, '1', or 'many',
     * }
     */
    cardinalityOfRelationship(relationshipName) {
        const relationship = this.relationships.filter(r => r.name === relationshipName)[0];

        const cardinality = {
            from: null,
            to: null,
        };

            
        if (relationship.singular) {
            cardinality.to = '1';
        }
        else {
            cardinality.to = 'many';
        }
            
        if (relationship.mirrorRelationship !== undefined) {
            const mirrorRelationship = AllClassModels[relationship.toClass].relationships.filter(x => x.name === relationship.mirrorRelationship)[0];


            if (mirrorRelationship.singular) {
                cardinality.from = '1';
            }
            else {
                cardinality.from = 'many';
            }
        }

        return cardinality;
    }

    /*
     * discriminated()
     * Determines if this ClassModel is discriminated. This ClassModel is considered discriminated if 
     *    it has a direct sub-ClassModel with its 'useSuperClassCollection' property set to true.
     * Returns
     * - Boolean - True if at least one of this ClassModel's direct sub-ClassModels has useSuperClassCollection 
     *    property equal to true. False otherwise.
     */
    discriminated() {
        for (const subClass of this.subClasses) {
            if (subClass.useSuperClassCollection)
                return true;
        }
        return false;
    }

    /*
     * firstNonNullPromiseResolution(promises)
     * Loops through given promises one at a time and returns the first non null resolution. 
     *    Will break the loop on the first non-null resolution. If none of the promises return 
     *    a non-null value, null is returned.
     * Parameters
     * - promises - Array<Promise> - An array of promises to wait for.
     * Returns 
     * - Promise<Any> - The resolved value of the first promise to resolve with a non-null value. Returns null
     *    if all promises resolve to null.
     * Throws
     * - Error - If any of the given promises reject with an error before another promise
     *    resolves with a non-null promise.
     */
    static async firstNonNullPromiseResolution(promises) {
        for (var index in promises) {
            let foundInstance = await promises[index];

            if (foundInstance != null) {
                return foundInstance;
                break;
            }
            else if (index == promises.length - 1) {
                return null;
            }
        }
    }

    /*
     * allPromiseResolutionsInstanceSets(promises)
     * Loops through and waits for the given promises one at a time. Each promise is expected to resolve
     *    to an InstanceSet. The promise resolutions are combined into a single InstanceSet with ClassModel
     *    of this ClassModel.
     * Parameters
     * - promises - Array<Promise> - An array of promises to wait for.
     * Returns 
     * - Promise<InstanceSet> - A Promise which resolves to an InstanceSet containing all the Instances
     *    of each InstanceSet that each of the given promises resolves with.
     * Throws
     * - Error - If any of the given promises reject with an error.
     */
    async allPromiseResoltionsInstanceSets(promises) {
        let results = new InstanceSet(this);

        for (var promise of promises) {
            let singleResult = await promise;
            if (!singleResult.isEmpty())
                results.addInstances(singleResult);
        }

        return results;
    }

    /* 
     * asyncFilter(instances, asyncFilterFunction)
     * A function which can filter an array of instances using an asynchronus function.
     * Parameters
     * - instances - Iterable<Instance> - An iterable (InstanceSet, Array, etc.) of Instances to filter.
     * - asyncFilterFunction - Function - An asynchronous function which accepts a single Instance as
     *    and argument and which will resolve true or false.
     * Returns
     * - Promise<Instances> - The Instances for which the given asyncFilter resolves true.
     */
    static async asyncFilter(instances, asyncFilterFunction) {
        let filtered = [];
        let filterPromises = [];

        instances.forEach((instance) => {
            filterPromises.push(asyncFilterFunction(instance));
        });

        let instanceIndex;

        for (instanceIndex = 0; instanceIndex < instances.length; instanceIndex++)  
            if (await filterPromises[instanceIndex])
                filtered.push(instances[instanceIndex]);
            
        return filtered;
    }

    /* 
     * isSuperClass()
     * Use to determine if this ClassModel is a super ClassModel.
     * Returns
     * - Boolean - True if this ClassModel has any sub ClassModels, false otherwise.
     */
    isSuperClass() {
        return (this.subClasses && this.subClasses.length)
    }

    /*
     * allSuperClasses()
     * Finds all the ClassModels which are a parent ClassModel to this ClassModel, 
     *    all the way up the inheritance tree.
     * Returns
     * - Array<ClassModel> - An array containing all the ClassModels that are a super ClassModel
     *    to this ClassModel.
     */
    allSuperClasses() {
        const superClasses = new SuperSet(this.superClasses);

        for (const superClass of this.superClasses) {
            superClasses.addFromIterable(superClass.allSuperClasses())
        }

        return [...superClasses];
    }

    /*
     * allSubClasses()
     * Finds all the ClassModels which are a child ClassModel to this ClassModel, 
     *    all the way down the inheritance tree.
     * Returns
     * - Array<ClassModel> - An array containing all the ClassModels that are a sub ClassModel
     *    of this ClassModel.
     */
    allSubClasses() {
        const subClasses = new SuperSet(this.subClasses);

        for (const subClass of this.subClasses) {
            subClasses.addFromIterable(subClass.allSubClasses())
        }

        return [...subClasses];
    }

    /*
     * emptyInstanceSet()
     * Creates a new empty InstanceSet of this classModel.
     * Returns
     * - An empty InstanceSet of this ClassModel.
     */
    emptyInstanceSet() {
        return new InstanceSet(this);
    }


    // Insert, Update, Delete Methods

    /*
     * insertOne(document)
     * Inserts a document into the collection for this ClassModel. Internal use only. 
     * Parameters
     * - document - Object - An object to insert into the collection for this ClassModel.
     * Returns
     * - Promise<insertOneWriteOpResult> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~insertOneWriteOpResult
     */
    async insertOne(document) {
        return database.insertOne(this.collection, document);
    }

    /*
     * insertMany(documents)
     * Inserts multiple documents into the collection for this ClassModel. Internal use only. 
     * Parameters
     * - documents - Array<Object> - An array of objects to insert into the collection for this ClassModel.
     * Returns
     * - Promise<insertWriteOpResultObject> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~insertWriteOpResult
     */
    async insertMany(documents) {
        return database.insertMany(this.collection, documents);
    }

    /* 
     * update(instance)
     * Updates the given instance in this ClassModel's collection to match the current state of the instance.
     *    Internal use only.
     * Parameters
     * - instance - Instance - An instance of this ClassModel to update in the database.
     * Returns
     * - Promise<updateWriteOpResult> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~updateWriteOpResult
     */
    async update(instance) {
        return database.update(this.collection, instance);
    }

    /* 
     * overwrite(instance)
     * Overwrites an instance in the collection for this ClassModel. Do Not Use. Internal Use Only.
     * Parameters
     * - instance - Instance - An instance of this ClassModel to overwrite in the database.
     * Returns
     * - Promise<updateWriteOpResult> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~updateWriteOpResult
     */
    async overwrite(instance) {
        return database.overwrite(this.collection, instance);
    }

    /*
     * delete(instance)
     * Deletes an instance from the collection for this ClassModel.
     * Parameters
     * - instance - Instance - An instance to delete.
     * Returns
     * - Promise<deleteWriteOpResult> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~deleteWriteOpResult
     * Throws
     * - NoommanArgumentError - If the given Instance's ClassModel is not this ClassModel.
     */
    async delete(instance) {

        if (instance.classModel !== this)
            throw new NoommanErrors.NoommanArgumentError(this.className + '.delete() called on an instance of a different class.');

        return database.deleteOne(this.collection, instance);
    }

    // Query Methods

    /* 
     * find(queryFilter, readControlMethodParameters, sensitiveControlMethodParameters)
     * Finds Instances of this ClassModel using the given query filter in the database. 
     *    If called on a super-ClassModel, will recursively check this ClassModel's collection, and then it's sub-ClassModels' collections.
     *    This method respects readControl and sensitiveControl methods. If this ClassModel is read controlled, Instances found
     *    during query will be filtered down to those which pass the readControl method(s) for this ClassModel. If this ClassModel
     *    is sensitive controlled, all Instances which do not pass the sensitiveControl method(s) for this ClassModel will be 
     *    stripped of any sensitive attributes.
     * Parameters
     * - queryFilter - Object - A mongo query object (required).
     * - readControlMethodParameters - Object - An object containing parameters that will be passed to the readControl method(s)
     *    for this ClassModel.
     * - sensitiveControlMethodParameters - Object - An object containing parameters that will be passed to the sensitiveControl
     *    method(s) for this ClassModel.
     * Returns 
     * - Promise<InstanceSet> - An InstanceSet of this ClassModel containing all instances of this ClassModel or its children
     *    which match the given query and pass the readControl methods if applicable.
     */
    async find(queryFilter, readControlMethodParameters, sensitiveControlMethodParameters) {
        const unfiltered = await this.pureFind(queryFilter);
        const filtered = await this.readControlFilter(unfiltered, readControlMethodParameters);
        await filtered.sensitiveControlCheckAndStrip(sensitiveControlMethodParameters);

        return filtered;
    }

    /* 
     * pureFind(queryFilter)
     * Finds instances of this ClassModel using the given query filter in the database. 
     *    If called on a super-ClassModel, will recursively check this ClassModel's collection, and then it's sub-ClassModels collections.
     *    This method DOES NOT do any readControl or sensitiveControl filtering. 
     * Parameters
     * - queryFilter - Object - A mongo query object (required).
     * Returns 
     * - Promise<InstanceSet> - An InstanceSet of this ClassModel containing all instances of this ClassModel or its children
     *    which match the given query.
     * Throws
     * - NoommanClassModelError - If this ClassModel is abstract and has no sub-ClassModels.
     */
    async pureFind(queryFilter={}) {
        const foundInstances = new InstanceSet(this);

        const subClassesWithDifferentCollections = this.subClasses ? this.subClasses.filter(subClass => !subClass.useSuperClassCollection) : [];

        // If this class is a non-discriminated abstract class and it doesn't have any sub classes, throw an error.
        if (this.abstract && !this.isSuperClass())
            throw new NoommanErrors.NoommanClassModelError('Error in ' + this.className + '.find(). This class is abstract and non-discriminated, but it has no sub-classes.');

        if (this.useSuperClassCollection) {
            queryFilter.__t = this.className;
        }

        if (this.collection) {
            const documentsFoundInThisCollection = await database.find(this.collection, queryFilter); 
            const instancesFoundInThisCollection = new InstanceSet(this, documentsFoundInThisCollection.map(document => { 
                if (document.__t)
                    return new Instance(AllClassModels[document.__t], document);
                return new Instance(this, document);
            }));
            foundInstances.addInstances(instancesFoundInThisCollection);
        }
        
        const promises = [];
  
        for (const subClass of subClassesWithDifferentCollections) {
            delete queryFilter.__t;
            promises.push(subClass.pureFind(queryFilter));
        }

        const instancesFoundOfSubClasses = await this.allPromiseResoltionsInstanceSets(promises);
        
        foundInstances.addInstances(instancesFoundOfSubClasses)
        return foundInstances;
    }

    /* 
     * findOne(queryFilter, readControlMethodParameters, sensitiveControlMethodParameters)
     * Finds a single Instance of this ClassModel using the given query filter in the database. 
     *    If called on a super-ClassModel, will recursively check this ClassModel's collection, and then it's sub-ClassModels collections.
     *    This method respects readControl and sensitiveControl methods. If this ClassModel is read controlled, the Instance found
     *    during query will not be returned if it does not pass the readControl method(s) for this ClassModel. If this ClassModel
     *    is sensitive controlled, an Instance which does not pass the sensitiveControl method(s) for this ClassModel will be 
     *    stripped of any sensitive attributes.
     * Parameters
     * - queryFilter - Object - A mongo query object (required).
     * - readControlMethodParameters - Object - An object containing parameters that will be passed to the readControl method(s)
     *    for this ClassModel.
     * - sensitiveControlMethodParameters - Object - An object containing parameters that will be passed to the sensitiveControl
     *    method(s) for this ClassModel.
     * Returns 
     * - Promise<Instance> - The first Instance of this ClassModel or its children which matches the given query and passes the
     *    readControl methods if applicable. Returns null if no Instance matches query or if matching Instance does not pass 
     *    readControl method if applicable.
     */
    async findOne(queryFilter, readControlMethodParameters, sensitiveControlMethodParameters) {
        const unfiltered = await this.pureFindOne(queryFilter);
        if (unfiltered === null) {
            return null;
        }

        const filtered = await this.readControlFilterInstance(unfiltered, readControlMethodParameters);
        if (filtered === null) {
            return null;
        }

        const needToStrip = await this.sensitiveControlFilterInstance(filtered, sensitiveControlMethodParameters);
        if (needToStrip !== null) {
            filtered.stripSensitiveAttributes();
        }

        return filtered;
    }

    /* 
     * pureFindOne(queryFilter)
     * Finds a single instance of this ClassModel using the given query filter in the database. 
     *    If called on a superclass, will recursively check this ClassModel's collection, and then it's sub-ClassModels collections.
     *    This method does not respect readControl and sensitiveControl methods. 
     * Parameters
     * - queryFilter - Object - A mongo query object (required).
     * Returns 
     * - Promise<Instance> - The first Instance of this ClassModel or its children which matches the given query.
     * Throws
     * - NoommanClassModelError - If this ClassModel is abstract and has no sub-ClassModels.
     */
    async pureFindOne(queryFilter) {
        const subClassesWithDifferentCollections = this.subClasses ? this.subClasses.filter(subClass => !subClass.useSuperClassCollection) : [];

        // If this class is a non-discriminated abstract class and it doesn't have any sub classes, throw an error.
        if (this.abstract && !this.isSuperClass())
            throw new NoommanErrors.NoommanClassModelError('Error in ' + this.className + '.findOne(). This class is abstract and non-discriminated, but it has no sub-classes.');

        if (this.useSuperClassCollection) {
            queryFilter.__t = this.className;
        }

        if (this.collection) {
            const documentFoundInThisCollection = await database.findOne(this.collection, queryFilter);

            if (documentFoundInThisCollection !== null) {
                if (documentFoundInThisCollection.__t)
                    return new Instance(AllClassModels[documentFoundInThisCollection.__t], documentFoundInThisCollection);
                else 
                    return new Instance(this, documentFoundInThisCollection);
            }

            if (subClassesWithDifferentCollections.length == 0)
                return null;
        }

        delete queryFilter.__t;

        const promises = [];

        // Call findOne on our subclasses as well.
        for (let subClass of subClassesWithDifferentCollections)
            promises.push(subClass.pureFindOne(queryFilter));

        return ClassModel.firstNonNullPromiseResolution(promises);
    }

    /* 
     * findById(queryFilter, readControlMethodParameters, sensitiveControlMethodParameters)
     * Finds a single instance of this ClassModel with the given id. 
     *    If called on a superclass, will recursively check this ClassModel's collection, and then it's sub-ClassModels collections.
     *    This method respects readControl and sensitiveControl methods. If this ClassModel is read controlled, the Instance found
     *    during query will not be returned if it does not pass the readControl method(s) for this ClassModel. If this ClassModel
     *    is sensitive controlled, an instance which does not pass the sensitiveControl method(s) for this ClassModel will be 
     *    stripped of any sensitive attributes.
     * Parameters
     * - id - ObjectId - A mongo ObjectId of the Instance you which to find.
     * - readControlMethodParameters - Object - An object containing parameters that will be passed to the readControl method(s)
     *    for this ClassModel.
     * - sensitiveControlMethodParameters - Object - An object containing parameters that will be passed to the sensitiveControl
     *    method(s) for this ClassModel.
     * Returns 
     * - Promise<Instance> - The first Instance of this ClassModel or its children
     *    which has the given id and passes the readControl methods if applicable. Returns null if no Instance with the given id is 
     *    found, or if matching instance does not pass readControl method (if applicable).
     */
    async findById(id, readControlMethodParameters, sensitiveControlMethodParameters) {
        return this.findOne({_id: id}, readControlMethodParameters, sensitiveControlMethodParameters);
    }

    /* 
     * pureFindById(id)
     * Finds a single instance of this ClassModel with the given id.
     *    If called on a superclass, will recursively check this ClassModel's collection, and then it's sub-ClassModels collections.
     *    This method does not respect readControl and sensitiveControl methods. 
     * Parameters
     * - id - ObjectId - A mongo ObjectId of the Instance you which to find.
     * Returns 
     * - Promise<Instance> - The Instance of this ClassModel or its children with the given id.
     */
    async pureFindById(id) {
        return this.pureFindOne({_id: id});
    }

    async findPage(queryFilter={}, page=0, pageSize=100, orderBy={_id: 1}, readControlMethodParameters, sensitiveControlMethodParameters) {
        if (this.abstract && !this.isSuperClass())
            throw new NoommanErrors.NoommanClassModelError('Error in ' + this.className + '.findPage(). This class is abstract and non-discriminated, but it has no sub-classes.');

        if (page < 0 || pageSize <= 0) {
            throw new Error(this.className + '.findPage() called with negative page or pageSize value.');
        }

        queryFilter = queryFilter ? queryFilter : {};
        let documents = [];
        const cursors = await this.findPageRecursive(queryFilter, page, pageSize, orderBy);
       
        let index = 0;
        const startIndex = page * pageSize;
        const endIndex = ((page + 1) * pageSize) - 1;
        let documentsRemaining = pageSize;
        let totalNumberOfInstances = 0;

        for (const cursor of cursors) {
            totalNumberOfInstances += await cursor.cursor.count();
        }

        if (startIndex > totalNumberOfInstances) {
            return {
                instances: new InstanceSet(this),
                page,
                pageSize,
                hiddenInstances: 0, 
                totalNumberOfInstances,
            }
        }

        for (const cursor of cursors) {
            const cursorCount = await cursor.cursor.count();
            const cursorStart = index;
            const cursorEnd = index + cursorCount - 1;

            if (endIndex < cursorStart || startIndex > cursorEnd) {
                index += cursorCount;
                continue;
            }
            else {
                const skipValue = startIndex > cursorStart ? startIndex - cursorStart : 0;
                let limitValue = documentsRemaining;
                const documentsFromThisCursor = await cursor.cursor.skip(skipValue).limit(limitValue).toArray();

                for (const document of documentsFromThisCursor) {
                    documents.push({
                        document,
                        className: document.__t ? document.__t : cursor.className,
                    });
                }
            }

            documentsRemaining = pageSize - documents.length;
            index += cursorCount;

            if (documentsRemaining === 0 || index >= totalNumberOfInstances) {
                break;
            }
        }

        // convert documents to instances
        const instances = new InstanceSet(this);

        for (const document of documents) {
            const documentClassModel = AllClassModels[document.className];
            instances.add(new Instance(documentClassModel, document.document));
        }

        const filteredInstances = await instances.readControlFilter(readControlMethodParameters);
        await filteredInstances.sensitiveControlCheckAndStrip(sensitiveControlMethodParameters);

        const hiddenInstances = instances.size - filteredInstances.size;

        return {
            instances: filteredInstances,
            page,
            pageSize,
            hiddenInstances, 
            totalNumberOfInstances,
        }
    }

    async findPageRecursive(queryFilter, page, pageSize, orderBy) {
        let cursorsWithClassName = [];

        const subClassesWithDifferentCollections = this.subClasses ? this.subClasses.filter(subClass => !subClass.useSuperClassCollection) : [];

        if (this.useSuperClassCollection) {
            queryFilter.__t = this.className;
        }

        if (this.collection) {
            const cursorForThisCollection = await database.findCursor(this.collection, queryFilter); 
            if(orderBy !== null) {
                cursorForThisCollection.sort(orderBy);
            }

            cursorsWithClassName.push( {
                className: this.className,
                cursor: cursorForThisCollection,
            });
        }
        
        const promises = [];
  
        for (const subClass of subClassesWithDifferentCollections) {
            delete queryFilter.__t;
            promises.push(subClass.findPageRecursive(queryFilter, page, pageSize, orderBy));
        }

        if (promises.length) {
            const subClassCursors = await Promise.all(promises);
            for (const subClassCursor of subClassCursors) {
                cursorsWithClassName = cursorsWithClassName.concat(subClassCursor);
            }
        }

        return cursorsWithClassName;
    }

    /*
     * updateRelatedInstancesForInstance(instance) 
     * Analyzes the changes to two-way relationships for the given Instance to determine which related instances also
     *    need to be updated in order to maintain the consistency of the two-way relationships. Will make the updates
     *    to those related Instances as necessary and save the changes to those instances.
     *    Internal use only.
     * Parameters
     * - instance - Instance - An instance of this ClassModel to analyze and  update related instances for.
     * Returns
     * - Promise<Array<Instance>> - An array of all the related instances which were updated.
     * Throws
     * - NoommanSaveError - If InstanceSet.saveWithoutRelatedUpdates() throws a NoommanSaveError.
     * - NoommanValidationError - If InstanceSet.saveWithoutRelatedUpdates() throws a NoommanValidationError.
     */
    async updateRelatedInstancesForInstance(instance) {
        const relatedDiff = instance.relatedDiffs();
        const reducedRelatedDiff = instance.reducedRelatedDiffs(relatedDiff);
        const instanceSet = new InstanceSet(NoommanClassModel);

        // Retrieve all related instances and collect them in an instanceSet.
        for (const relationshipName of Object.keys(relatedDiff)) {
            const relationship = this.relationships.filter(r => r.name === relationshipName)[0];
            const relatedInstances = await instance.walk(relationshipName);
            const previousRelatedInstances = await instance.walk(relationshipName, true);

            if (relationship.singular) {
                if (relatedInstances !== null) {
                    instanceSet.add(relatedInstances);
                }
                if (previousRelatedInstances !== null) {
                    instanceSet.add(previousRelatedInstances);
                }
            }
            else {
                if (relatedInstances instanceof InstanceSet && !relatedInstances.isEmpty()) {
                    instanceSet.addInstances(relatedInstances);
                }
                if (previousRelatedInstances instanceof InstanceSet && !previousRelatedInstances.isEmpty()) {
                    instanceSet.addInstances(previousRelatedInstances);
                }
            }

        }

        // Apply changes to related instances
        for (const id of Object.keys(reducedRelatedDiff)) {
            const relatedInstance = instanceSet.getInstanceWithId(id);
            relatedInstance.applyChanges(reducedRelatedDiff[id]);
        }

        return instanceSet.saveWithoutRelatedUpdates();
    }

    /*
     * updateRelatedInstancesForInstanceSet(instanceSet)
     * Analyzes the changes to two-way relationships for the Instances in the given InstanceSet to determine which 
     *    related Instances also need to be updated in order to maintain the consistency of the two-way relationships. 
     *    Will make the updates to those related instances as necessary and save the changes to those instances.
     *    Internal use only.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to analyze and update related instances for.
     * Returns
     * - Promise<Array<Instance>> - An array of all the related instances which were updated.
     * Throws
     * - NoommanSaveError - If InstanceSet.saveWithoutRelatedUpdates() throws a NoommanSaveError.
     * - NoommanValidationError - If InstanceSet.saveWithoutRelatedUpdates() throws a NoommanValidationError.
     */
    async updateRelatedInstancesForInstanceSet(instanceSet) {
        const relationshipsNeedingUpdate = new SuperSet();
        const relatedDiffs = [];

        if (!(instanceSet instanceof InstanceSet) || instanceSet.isEmpty()) {
            return;
        }

        for (const instance of instanceSet) {
            const relatedDiff = instance.relatedDiffs();
            const relationshipsToUpdateForInstance = Object.keys(relatedDiff);
            for (const relationshipName of relationshipsToUpdateForInstance) {
                relationshipsNeedingUpdate.add(relationshipName);
            }
            if (relationshipsToUpdateForInstance.length) {
                relatedDiffs.push(relatedDiff);
            }
        }

        if (relationshipsNeedingUpdate.isEmpty()) {
            return;
        }

        const combinedDiff = Diffable.combineMultipleRelatedDiffs(relatedDiffs);
        const allRelatedInstances = new InstanceSet(NoommanClassModel);

        // Retrieve all related instances and collect them in an instanceSet.
        for (const relationshipName of relationshipsNeedingUpdate) {
            const relatedInstances = await instanceSet.walk(relationshipName);
            const previousRelatedInstances = await instanceSet.walk(relationshipName, true);

            if (relatedInstances instanceof InstanceSet && !relatedInstances.isEmpty()) {
                allRelatedInstances.addInstances(relatedInstances);
            }
            if (previousRelatedInstances instanceof InstanceSet && !previousRelatedInstances.isEmpty()) {
                allRelatedInstances.addInstances(previousRelatedInstances);
            }
        }

        // Apply changes to related instances
        for (const id of Object.keys(combinedDiff)) {
            const relatedInstance = allRelatedInstances.getInstanceWithId(id);
            relatedInstance.applyChanges(combinedDiff[id]);
        }

        return allRelatedInstances.saveWithoutRelatedUpdates();
    }

    // Crud Control Methods

    /*
     * evaluateCrudControlMethods(instanceSet, controlMethods, methodParameters)
     * Runs the crudControl methods determined by the controlMethods parameter for each Instance in the 
     *    given InstanceSet that are applicable for each Instance's ClassModel, using the given methodParameters.
     *    Returns an InstanceSet of Instances which for which at least one controlMethod returns false.
     *    Internal use only.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate crudControl methods on.
     * - controlMethods - String - A string which determines which type of control methods to run. Valid values
     *    are 'readControlMethods', 'createControlMethods', 'updateControlMethods', 'deleteControlMethods',
     *    or 'sensitiveControlMethods'.
     * - methodParameters - Object - An object containing any parameters that a particular crudControl method
     *    may need.
     * Returns
     * - Promise<InstanceSet> - An InstanceSet containing all the instances for which at least one crudControl
     *    method returns false.
     */
    async evaluateCrudControlMethods(instanceSet, controlMethods, methodParameters) {
        let rejectedInstances = new InstanceSet(this);

        const instancesOfThisClass = instanceSet.filterToInstanceSet(instance => {
            return instance.classModel === this;
        });

        for (const instance of instancesOfThisClass) {
            for (const controlMethod of this[controlMethods]) {
                let result = controlMethod.call(instance, methodParameters);
                if (result instanceof Promise) {
                    result = await result;
                }
                if (!result) {
                    rejectedInstances.add(instance);
                    continue;
                }
            }
        }

        if (this.isSuperClass()) {
            for (let subClass of this.subClasses) {
                let instancesOfSubClass = instanceSet.filterForClassModel(subClass);

                if (!instancesOfSubClass.isEmpty()) {
                    const rejectedSubClassInstances = await subClass.evaluateCrudControlMethods(instancesOfSubClass, controlMethods, methodParameters);
                    rejectedInstances.addFromIterable(rejectedSubClassInstances);
                }
            }
        }

        return rejectedInstances;
    }

    /*
     * createControlCheck(instanceSet, createControlMethodParameters)
     * Runs applicable createControl methods for each Instance in the given InstanceSet, and throws an error if any
     *    createControl method returns false for any of the Instances in the InstanceSet.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate createControl methods on.
     * - createControlMethodParameters - Object - An object containing any parameters that the createControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if all Instances pass the createControl methods.
     * Throws
     * - NoommanArgumentError - If instanceSet parameter is not an InstanceSet.
     * - NoommanSaveError - If any createControl method returns false for any of the Instances in the InstanceSet.
     */
    async createControlCheck(instanceSet, createControlMethodParameters) {
        if (!(instanceSet instanceof InstanceSet))
            throw new NoommanErrors.NoommanArgumentError('Incorrect parameters. ' + this.className + '.createControlCheck(InstanceSet instanceSet, createControlMethodParameters)');
        
        if (instanceSet.isEmpty() || !this.createControlMethods.length)
            return;

        const rejectedInstances = await this.evaluateCrudControlMethods(instanceSet, 'createControlMethods', createControlMethodParameters);

        if (!rejectedInstances.isEmpty())
            throw new NoommanErrors.NoommanSaveError('Illegal attempt to create instances: ' + rejectedInstances.getInstanceIds());
    }


    /*
     * createControlCheckInstance(instance, createControlMethodParameters)
     * Runs applicable createControl methods for the given Instance, and throws an error if any
     *    createControl method returns false for the Instance.
     * Parameters
     * - instance - Instance - An Instance of this ClassModel to evaluate createControl methods on.
     * - createControlMethodParameters - Object - An object containing any parameters that the createControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if all Instances pass the createControl methods.
     * Throws
     * - NoommanArgumentError - If instance is not an Instance of this ClassModel.
     * - NoommanSaveError - If any createControl method returns false for the given Instance.
     */
    async createControlCheckInstance(instance, createControlMethodParameters) {
        const instanceSet = new InstanceSet(instance.classModel, [instance]);
        return this.createControlCheck(instanceSet, createControlMethodParameters);
    }

    /*
     * readControlFilter(instanceSet, readControlMethodParameters)
     * Runs applicable readControl methods for each Instance in the given InstanceSet, and filters out any
     *    Instances for which any readControl method returns false.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate readControl methods on.
     * - readControlMethodParameters - Object - An object containing any parameters that the readControl method(s)
     *    may need.
     * Returns
     * - Promise<InstanceSet> - An InstanceSet containing those Instances for which all readControl methods return true.
     * Throws
     * - NoommanArgumentError - If instanceSet parameter is not an InstanceSet.
     */
    async readControlFilter(instanceSet, readControlMethodParameters) {
        if (!(instanceSet instanceof InstanceSet))
            throw new NoommanErrors.NoommanArgumentError('Incorrect parameters. ' + this.className + '.readControlFilter(InstanceSet instanceSet, readControlMethodParameters)');
        
        // If InstanceSet is empty or not read controlled, just return a copy of it.
        if (!instanceSet.size || this.readControlMethods.length === 0)
            return new InstanceSet(this, instanceSet);

        const rejectedInstances = await this.evaluateCrudControlMethods(instanceSet, 'readControlMethods', readControlMethodParameters);

        return instanceSet.difference(rejectedInstances);
    }

    /*
     * readControlFilterInstance(instance, readControlMethodParameters)
     * Runs applicable readControl methods for the given Instance. If each readControl method returns true for 
     *    the Instance, then the Instance is returned, otherwise null is returned.
     * Parameters
     * - instance - Instance - An Instance of this ClassModel to evaluate readControl methods on.
     * - readControlMethodParameters - Object - An object containing any parameters that the readControl method(s)
     *    may need.
     * Returns
     * - Promise<Instance> - The given Instance if all readControl methods return true, otherwise null.
     * Throws
     * - NoommanArgumentError - If instance is not an Instance of this ClassModel.
     */
    async readControlFilterInstance(instance, readControlMethodParameters) {
        const instanceSet = new InstanceSet(this, [instance]);
        const filteredInstanceSet = await this.readControlFilter(instanceSet, readControlMethodParameters);
        return filteredInstanceSet.isEmpty() ? null : [...instanceSet][0];
    }

    /*
     * updateControlCheck(instanceSet, updateControlMethodParameters)
     * Runs applicable updateControl methods for each Instance in the given InstanceSet, throws an error 
     *    if any updateControl method returns false for any Instance.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate updateControl methods on.
     * - updateControlMethodParameters - Object - An object containing any parameters that the updateControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if all Instances pass the updateControl methods.
     * Throws
     * - NoommanArgumentError - If instanceSet parameter is not an InstanceSet.
     * - NoommanSaveError - If any updateControl method returns false for any of the Instances in the InstanceSet.
     */
    async updateControlCheck(instanceSet, updateControlMethodParameters) {
        if (!(instanceSet instanceof InstanceSet))
            throw new NoommanErrors.NoommanArgumentError('Incorrect parameters. ' + this.className + '.updateControlCheck(InstanceSet instanceSet, updateControlMethodParameters)');

        if (instanceSet.isEmpty() || !this.updateControlMethods.length)
            return;

        const rejectedInstances = await this.evaluateCrudControlMethods(instanceSet, 'updateControlMethods', updateControlMethodParameters);

        if (!rejectedInstances.isEmpty())
            throw new NoommanErrors.NoommanSaveError('Illegal attempt to update instances: ' + rejectedInstances.getInstanceIds());
    }

    /*
     * updateControlCheckInstance(instance, updateControlMethodParameters)
     * Runs applicable updateControl methods for the given Instance, throws an error 
     *    if any updateControl method returns false for the Instance.
     * Parameters
     * - instance - Instance - An Instance of this ClassModel to evaluate updateControl methods on.
     * - updateControlMethodParameters - Object - An object containing any parameters that the updateControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if the Instance passes the updateControl methods.
     * Throws
     * - NoommanArgumentError - If instance is not an Instance of this ClassModel.
     * - NoommanSaveError - If any updateControl method returns false for the given Instance.
     */
    async updateControlCheckInstance(instance, updateControlMethodParameters) {
        const instanceSet = new InstanceSet(instance.classModel, [instance]);
        return this.updateControlCheck(instanceSet, updateControlMethodParameters);
    }

    /*
     * deleteControlCheck(instanceSet, deleteControlMethodParameters)
     * Runs applicable deleteControl methods for each Instance in the given InstanceSet, throws an error 
     *    if any deleteControl method returns false for any Instance.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate deleteControl methods on.
     * - deleteControlMethodParameters - Object - An object containing any parameters that the deleteControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if all Instances pass the deleteControl methods.
     * Throws
     * - NoommanArgumentError - If instanceSet parameter is not an InstanceSet.
     * - NoommanDeleteError - If any deleteControl method returns false for any of the Instances in the InstanceSet.
     */
    async deleteControlCheck(instanceSet, deleteControlMethodParameters) {
        if (!(instanceSet instanceof InstanceSet))
            throw new NoommanErrors.NoommanArgumentError('Incorrect parameters. ' + this.className + '.deleteControlCheck(InstanceSet instanceSet, deleteControlMethodParameters)');

        if (instanceSet.isEmpty() || !this.deleteControlMethods.length)
            return;

        const rejectedInstances = await this.evaluateCrudControlMethods(instanceSet, 'deleteControlMethods', deleteControlMethodParameters);

        if (!rejectedInstances.isEmpty())
            throw new NoommanErrors.NoommanDeleteError('Illegal attempt to delete instances: ' + rejectedInstances.getInstanceIds());
    }

    /*
     * deleteControlCheckInstance(instance, deleteControlMethodParameters)
     * Runs applicable deleteControl methods for the given Instance, throws an error 
     *    if any deleteControl method returns false for the Instance.
     * Parameters
     * - instance - Instance - An Instance of this ClassModel to evaluate deleteControl methods on.
     * - deleteControlMethodParameters - Object - An object containing any parameters that the deleteControl method(s)
     *    may need.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if the Instance passes the deleteControl methods.
     * Throws
     * - NoommanArgumentError - If instance is not an Instance of this ClassModel.
     * - NoommanDeleteError - If any deleteControl method returns false for the given Instance.
     */
    async deleteControlCheckInstance(instance, deleteControlMethodParameters) {
        const instanceSet = new InstanceSet(instance.classModel, [instance]);
        return this.deleteControlCheck(instanceSet, deleteControlMethodParameters);
    }

    /*
     * sensitiveControlFilter(instanceSet, sensitiveControlMethodParameters)
     * Runs applicable sensitiveControl methods for each Instance in the given InstanceSet, and returns those for which
     *    at least one sensitiveControl method fails.
     * Parameters
     * - instanceSet - InstanceSet - An InstanceSet of this ClassModel to evaluate sensitiveControl methods on.
     * - sensitiveControlMethodParameters - Object - An object containing any parameters that the sensitiveControl method(s)
     *    may need.
     * Returns
     * - Promise<InstanceSet> - An InstanceSet containing those Instances for which any sensitiveControl method returns false.
     * Throws
     * - NoommanArgumentError - If instanceSet parameter is not an InstanceSet.
     */
    async sensitiveControlFilter(instanceSet, sensitiveControlMethodParameters) {
        if (!(instanceSet instanceof InstanceSet))
            throw new NoommanErrors.NoommanArgumentError('Incorrect parameters. ' + this.className + '.sensitiveControlFilter(InstanceSet instanceSet, sensitiveControlMethodParameters)');
        
        // If InstanceSet is empty or not sensitive controlled, just return a copy of it.
        if (!instanceSet.size || this.sensitiveControlMethods.length === 0)
            return new InstanceSet(this, instanceSet);

        const rejectedInstances = await this.evaluateCrudControlMethods(instanceSet, 'sensitiveControlMethods', sensitiveControlMethodParameters);

        return rejectedInstances;
    }

    /*
     * sensitiveControlFilterInstance(instance, sensitiveControlMethodParameters)
     * Runs applicable sensitiveControl methods for the given Instance, and returns the instance if 
     *    any sensitiveControl method returns false. Returns null otherwise.
     * Parameters
     * - instance - Instance - An Instance of this ClassModel to evaluate sensitiveControl methods on.
     * - sensitiveControlMethodParameters - Object - An object containing any parameters that the sensitiveControl method(s)
     *    may need.
     * Returns
     * - Promise<Instance> - The given Instance if any sensitiveControl method returns false, otherwise null.
     * Throws
     * - NoommanArgumentError - If instance is not an Instance of this ClassModel.
     */
    async sensitiveControlFilterInstance(instance, sensitiveControlMethodParameters) {
        const instanceSet = new InstanceSet(this, [instance]);
        const rejectedInstances = await this.sensitiveControlFilter(instanceSet, sensitiveControlMethodParameters);
        return rejectedInstances.size > 0 ? instance : null;
    }

    /*
     * deleteMany(instances)
     * Deletes all of the given instances from this ClassModel's collection.
     * Parameters
     * - instances - Iterable<Instance> - An Iterable (Array, InstanceSet, etc.) containing Instances to delete.
     * Returns 
     * - Promise<deleteWriteOpResult> - See https://mongodb.github.io/node-mongodb-native/3.3/api/Collection.html#~deleteWriteOpResult
     */
    async deleteMany(instances) {
        return database.deleteMany(this.collection, instances);
    }

    /* 
     * clear() 
     * Deletes every document in the collection for this ClassModel. 
     *    This is for testing purposes only, never run in production.
     * Returns
     * - Promise<undefined> - A promise which resolves to undefined if no Errors are thrown.
     * Throws
     * - NoommanClassModelError - If this is an abstract, non-discriminated ClassModel.
     */
    async clear() {
        if (this.abstract && !this.discriminated())
            throw new NoommanErrors.NoommanClassModelError('Cannot call clear() on an abstract, non-discriminated class. Class: ' + classModel.className);

        if (this.useSuperClassCollection) {
            return database.collection(this.collection).deleteMany({ __t: this.className });
        }
        else {
            return database.collection(this.collection).deleteMany({});
        }        
    }
}

/* 
 * NoommanClassModel
 * An internal ClassModel which is the root of the ClassModel inheritance tree. Every
 *    ClassModel created will be considered a sub-ClassModel of the NoommanClassModel.
 */
const NoommanClassModel = new ClassModel({
    className: 'NoommanClassModel',
    abstract: true,
});

module.exports = ClassModel;
